compiler/rustc_hir_typeck/src/pat.rs

```diff
@@ -1881,17 +1891,18 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
         let len = unmentioned_fields.len();
         let (prefix, postfix, sp) = match fields {
             [] => match &pat.kind {
-                PatKind::Struct(path, [], false) => {
+                PatKind::Struct(path, [], false) if path.span().eq_ctxt(pat.span) => {
                     (" { ", " }", path.span().shrink_to_hi().until(pat.span.shrink_to_hi()))
                 }
                 _ => return err,
             },
             [.., field] => {
-                // Account for last field having a trailing comma or parse recovery at the tail of
-                // the pattern to avoid invalid suggestion (#78511).
-                let tail = field.span.shrink_to_hi().with_hi(pat.span.hi());
                 match &pat.kind {
-                    PatKind::Struct(..) => (", ", " }", tail),
+                    // Account for last field having a trailing comma or parse recovery at the tail of
+                    // the pattern to avoid invalid suggestion (#78511).
+                    PatKind::Struct(..) if field.span.eq_ctxt(pat.span) => {
+                        (", ", " }", field.span.shrink_to_hi().with_hi(pat.span.hi()))
```

Comment:
You can generalize this by looking at `field.span.ancestor_within_span(pat.span)`.

Reply:
I've already tried using `find_ancestor_inside` in some earlier version of this patch before opening a PR without a satisfying result if I remember correctly. However, let me check again tomorrow, I might've missed something.

Reply:
Hmm, so I'm almost certain that `pat.span.FIND_ANCESTOR(field.span)` with `FIND_ANCESTOR` ∊ {`find_ancestor_inside`, `find_ancestor_in_same_ctxt`} or just `field.span` isn't what we want here, `eq_ctxt` is conservative and correct.

If the field and the struct patterns aren't in the same syntax context, there's not much we can do here structured-suggestion-wise without resorting to elaborate `span_to_snippet` hacks. If the field comes from a different expansion, then the struct pattern has to be of the form `S { … $field … }` or `S { … $( … )… … }`, meaning the field has to be a metavariable or a repetition but notably not a macro call like `S { … m!(…) … }` (ungrammatical). Therefore, there are two location were we could theoretically suggest the unmentioned fields:

1. At the end of the struct pattern (~`pat.span`): However, we don't know if we should add a leading comma or not since we'd need to look at def-site (consider `S { $( $fields/*,*/ )* }`) and use-site (consider `m!(field/*,*/)`) which requires `span_to_snippet` I think and is ugly. The suggestion is marked `MachineApplicable` btw (we could ofc tweak that in this case to be `MaybeIncorrect`).
2. At the end of the last field (~`field.span`)
  * However, the other fields (other than the last one) might come from a more "general" expansion.
  * The suggestion might be incorrect since we don't know the shape of the metavariable `$field` (etc.). Consider `m!(field: _)` where we don't know if the suggestion `m!(field: _, missing: _)` would be valid input for macro `m`

Reply:
I've noticed that `eq_ctxt` is not sufficient anyway for preventing invalid suggestions. For example:

```rs
struct S {
    a: (),
    b: (),
}

macro_rules! n {
    ($( $f:ident )?) => {
        let S { $( $f )? } = loop {};
    }
}

fn main() {
    n!(a);
}

```

Here, we suggest garbage:

```
error[E0027]: pattern does not mention field `b`
  --> krash.rs:17:13
   |
17 |         let S { $( $f )? } = loop {};
   |             ^^^^^^^^^^^^^^ missing field `b`
...
23 |     n!(a);
   |     ----- in this macro invocation
   |
   = note: this error originates in the macro `n` (in Nightly builds, run with -Z macro-backtrace for more info)
help: include the missing field in the pattern
   |
17 |         let S { $( $f, b } = loop {};
   |                      ~~~~~
help: if you don't care about this missing field, you can explicitly ignore it
   |
17 |         let S { $( $f, .. } = loop {};
   |                      ~~~~~~
```

On master, we don't but on the other hand it misclassifies `a` and `b` to be inaccessible:

```
error: pattern requires `..` due to inaccessible fields
  --> krash.rs:17:13
   |
17 |         let S { $( $f )? } = loop {};
   |             ^^^^^^^^^^^^^^
...
23 |     n!(a);
   |     ----- in this macro invocation
   |
   = note: this error originates in the macro `n` (in Nightly builds, run with -Z macro-backtrace for more info)
help: ignore the inaccessible and unused fields
   |
17 |         let S { $( $f, .. )? } = loop {};
   |                      ++++
```

Seems like I still need to tweak some things :|.